강원대 BE 이지혜 - Step1 (카카오 로그인)

xqqldir opened the pull request:
늦게 제출하여 죄송합니다.. ㅠㅠ

xqqldir committed:
[Docs] README Update - 구현 예정 기능

xqqldir committed:
[Feat] 인가 코드 & 토큰 받기

xqqldir committed:
[Feat] 카카오 API 연동 로그인

iphwade commented on src/main/java/gift/config/WebConfig.java:
```diff
@@ -18,6 +20,11 @@ public WebConfig(AuthInterceptor authInterceptor) {
-}
\ No newline at end of file
+
+    @Bean
+    public RestTemplate restTemplate() {
+        return new RestTemplate();
+    }
```
외부API 호출처가 하나 더 늘어났을 때 이 RestTemplate bean을 공유해야할까요?

xqqldir replied:
네, 공유를 해야 합니다. 
매번 새로 만들어 보낸다면 메모리 낭비도 심하고 일관되지 못하기 때문입니다.

iphwade replied:
네 메모리, 일관성적인 측면에서는 지혜님이 말씀하신 의견도 충분히 의미가 있습니다.
다만, 이런 형태로 구현하게 되면 RestTemplate을 통해 설정되는 모든 외부 API들이 일관적인 설정들을 가져야한다는 점인데요.

예를 들면, baseUrl을 설정하는 상황을 생각해보면, 네이버, 카카오 로그인을 구현해야해서 baseUrl이 달라지는 상황은 어떻게 구현해야할까요? 
또, 네이버는 1초의 타임아웃을 설정해야하고 카카오는 3초의 타임아웃을 설정해야하는 상황은 하나의 RestTemplate을 공유하며 사용하긴 어렵겠네요.

iphwade commented on src/main/java/gift/controller/KakaoController.java:
```diff
@@ -0,0 +1,45 @@
+    private String clientId;
+
+    @GetMapping("/")
+    public String kakaoCallback(@RequestParam String code) {
```
String 타입을 반환하는 것보다 적절한 DTO를 반환해보는 건 어떨까요?

xqqldir replied:
넵 반영해보겠습니다.

iphwade commented on src/main/java/gift/controller/KakaoController.java:
```diff
@@ -0,0 +1,45 @@
+            return "토큰 발급에 실패했습니다.";
+        }
+
+        return "액세스 토큰 발급 성공: " + tokenResponse.getAccessToken();
```
Controller에서 처리해야하는 범위가 너무 크군요. 
레이어를 적용해서 처리의 책임을 나눠볼까요? (예, Service로 분리, 인프라 로직을 따로 분리)

xqqldir replied:
Service로 분리하는 쪽으로 반영해보겠습니다!

iphwade commented on src/main/java/gift/controller/KakaoController.java:
```diff
@@ -0,0 +1,45 @@
+    @GetMapping("/")
+    public String kakaoCallback(@RequestParam String code) {
+        // 토큰 발급
+        String tokenUrl = "https://kauth.kakao.com/oauth/token";
```
RestTemplate과 같이 외부 API를 호출할 때 공통 설정들은 묶는다던지(예, baseUrl) 타임아웃을 설정한다던지 여러 설정들을 설정해줄 수 있는데요. 그런 것들을 고려해보면 어떨까요?

꼭 타임아웃만 있는 건 아니에요. 외부 API 요청할 때 생각해봐야할 것들을 학습해보면 좋을 것 같아요.

xqqldir replied:
넵 외부 API를 처음 써봐서 조금 더 공부해보겠습니당

iphwade commented on src/main/java/gift/controller/KakaoController.java:
```diff
@@ -0,0 +1,45 @@
+        // 토큰 발급
+        String tokenUrl = "https://kauth.kakao.com/oauth/token";
+        HttpHeaders tokenHeaders = new HttpHeaders();
+        tokenHeaders.setContentType(MediaType.APPLICATION_FORM_URLENCODED);
```
혹시 contentType은 정해져있었을까요-? 다른 type을 사용할 수 있을까해서요.

xqqldir replied:
넵 정해져있는 것 같습니다!

https://developers.kakao.com/docs/latest/ko/kakaologin/rest-api#request-token

iphwade commented on src/main/java/gift/controller/KakaoController.java:
```diff
@@ -0,0 +1,45 @@
+        HttpHeaders tokenHeaders = new HttpHeaders();
+        tokenHeaders.setContentType(MediaType.APPLICATION_FORM_URLENCODED);
+
+        MultiValueMap<String, String> tokenBody = new LinkedMultiValueMap<>();
+        tokenBody.add("grant_type", "authorization_code");
+        tokenBody.add("client_id", clientId);
+        tokenBody.add("redirect_uri", "http://localhost:8080");
+        tokenBody.add("code", code);
```
Request 객체를 만들어서 처리할 수 있을까요-?

xqqldir replied:
반영해보겠습니다.

iphwade commented on src/main/java/gift/dto/KakaoTokenResponse.java:
```diff
@@ -0,0 +1,12 @@
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class KakaoTokenResponse {
+    @JsonProperty("access_token")
```
👍

iphwade commented on src/main/java/gift/controller/KakaoController.java:
```diff
@@ -15,31 +17,41 @@
+        String accessToken = tokenResponse.getAccessToken();
+
+        // 사용자 정보 가져오기
+        String userInfoUrl = "https://kapi.kakao.com/v2/user/me";
```
https://kauth.kakao.com/가 공통적이네요. 합칠 수 있을까요?

xqqldir replied:
현재 코멘트 남겨주신 부분은 https://**kauth**.kakao.com/가 아닌 https://**kapi**.kakao.com/ 으로 보여집니다! 
제가 잘못 이해한 부분이 있다면 코멘트 남겨주시면 반영해보겠습니다!

iphwade replied:
엇! 그렇네요. 전 뭘 본걸까요 🤯

iphwade reviewed (changes requested):
수고하셨습니다-! 몇가지 코멘트 남겨두었는데요~ 답변/반영 해주세요-! :)

xqqldir committed:
[refactor] 피드백 반영

1. 기존의 kakaoController를 kakaoController와 KakaoService로 분리
2. kakaoCallback 함수에서 DTO 반환 (기존:String 반환)
3. DTO 반환을 위한 DTO 객체 생성
